src/wallet/walletdb.h

```diff
@@ -103,9 +104,16 @@ class CHDChain
         return accountCounter;
     }
 
-    void IncrementAccountCounter() {
-        // TODO: We should check for overflow somewhere and handle it.
-        accountCounter += 1;
+    /** Increments the account counter by 1 and returns it. Returns std::nullopt
+     *  if the increment operation would cause an overflow. */
+    std::optional<uint32_t> IncrementAccountCounter() {
+        auto newAccountCounter = accountCounter + 1;
+        if (newAccountCounter > (HARDENED_KEY_LIMIT - 1)) {
```

Comment:
We use `HARDENED_KEY_LIMIT - 1` as the "legacy account" for deriving legacy transparent and Sapling keys, so it's always in use. The last valid value for `accountCounter` is therefore `ZCASH_LEGACY_ACCOUNT - 1`.
```suggestion
        if (newAccountCounter >= ZCASH_LEGACY_ACCOUNT) {
```